### PR TITLE
Refactor CompilationStrategy and CompilationController

### PR DESCRIPTION
**Pull request by jamesgua**
- using CompilationController directly from OMR
- making CompilationStrategy extensible class

Signed-off-by: Tao Guan <james_mango@yahoo.com>

### CODE REVIEW

File: runtime/compiler/control/J9CompilationStrategy.hpp [hunk 1]
```diff
@@ -20,24 +20,32 @@
...
 class TR_PersistentMethodInfo;
-struct TR_MethodToBeCompiled;
-
-int32_t printTruncatedSignature(char *sigBuf, int32_t bufLen, J9Method *j9method);
+class TR_PersistentJittedBodyInfo;
 
-extern "C" {
-int32_t returnIprofilerState();
-}
+static int32_t interpreterProfilingState = IPROFILING_STATE_OFF;
```
Comment (mpirvu) [inline runtime/compiler/control/J9CompilationStrategy.hpp:48]: This defines a new variable. We want to use the one in HookedByTheJIT.cpp

File: runtime/compiler/control/J9CompilationStrategy.cpp [hunk 2]
```diff
@@ -39,95 +39,26 @@
...
-   return _useController;
-   }
-
-
-//-------------------------------- shutdown ---------------------------------
-// Called at shutdown time after compilation thread has been stopped
-// --------------------------------------------------------------------------
-void TR::CompilationController::shutdown()
+TR::CompilationStrategy*
+J9::CompilationStrategy::createInstance(const char* name)
```
Comment (mpirvu) [inline runtime/compiler/control/J9CompilationStrategy.cpp:43]: I would need to look at the OMR changes to see if this enough. Please add a link to the omr PR.

**Comment by 0xdaryl** [comment]
I believe this requires a coordinated merge with eclipse/omr#6884

File: runtime/compiler/control/J9CompilationStrategy.cpp [hunk 3]
```diff
@@ -39,95 +39,26 @@
...
-            if (verboseController)
-               setVerbose(atoi(verboseController));
-            if (verbose() >= LEVEL1)
-               fprintf(stderr, "Using %s comp strategy\n", strategyName);
-            }
-         }
-      }
-   //TR_ASSERT(_useController, "Must use compilation controller");
-//#ifdef COMP_YIELD_ANALYSIS
-   if (options->getOption(TR_EnableCompYieldStats))
```
Comment (mpirvu) [inline runtime/compiler/control/J9CompilationStrategy.cpp:90]: Is this done in some other part of the code? I don't see it in the omr linked PR.

Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.cpp:90]: Yes, dependent OMR change is https://github.com/eclipse/omr/pull/6884

File: runtime/compiler/control/J9CompilationStrategy.hpp (hunk 1 above)
Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.hpp:48]: It's mainly for compilation success:
Before - **interpreterProfilingState** was _local_ and _static_ to HookedByTheJIT.cpp, only way to get it is via method returnIprofilerState declared in **CompilationController.hpp**
After - **CompilationController.hpp** removed, new file J9CompilationStrategy.hpp added, in J9CompilationStrategy.cpp there's one place refer the value. 

I can keep using the getter as before if you prefer

File: runtime/compiler/control/J9CompilationStrategy.cpp (hunk 2 above)
Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.cpp:43]: dependent OMR change is https://github.com/eclipse/omr/pull/6884

File: runtime/compiler/control/J9CompilationStrategy.hpp (hunk 1 above)
Comment (mpirvu) [reply runtime/compiler/control/J9CompilationStrategy.hpp:48]: It's about correctness, not preference. Here's a link that explains the problems with static variables declared in headers: https://stackoverflow.com/questions/3837490/initializing-a-static-variable-in-header

File: runtime/compiler/control/J9CompilationStrategy.cpp [hunk 4]
```diff
@@ -39,94 +39,22 @@
...
-   int32_t remainingPlans = TR_OptimizationPlan::freeEntirePool();
-      // print some stats
-   if (verbose() >= LEVEL1)
-      {
-      fprintf(stderr, "Remaining optimizations plans in the system: %d\n", remainingPlans);
-      }
-   _compilationStrategy->shutdown();
+   J9::CompilationStrategy* instance = NULL;
+   instance = new (PERSISTENT_NEW) J9::CompilationStrategy();
+   return static_cast<TR::CompilationStrategy*>(instance);
```
Comment (mpirvu) [inline runtime/compiler/control/J9CompilationStrategy.cpp:47]: Would it work without the cast?

File: runtime/compiler/control/J9CompilationStrategy.cpp (hunk 3 above)
Comment (mpirvu) [reply runtime/compiler/control/J9CompilationStrategy.cpp:89]: I don't see where we allocate CompYieldStatsMatrix. Could you please point that out to me (filename and line number)?

Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.cpp:89]: I don't see either, only appear in J9Compilation [hpp](https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/compile/J9Compilation.hpp#L111) and [cpp](https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/compile/J9Compilation.cpp#L313)

File: runtime/compiler/control/J9CompilationStrategy.cpp (hunk 4 above)
Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.cpp:47]: Updated - remove static method createInstance now, add self method in OMRCompilationStrategy and use the method in J9CompilationStrategy

File: runtime/compiler/control/J9CompilationStrategy.hpp [hunk 5]
```diff
@@ -233,6 +237,10 @@ class DefaultCompilationStrategy : public TR::CompilationStrategy
    // statistics regarding the events it receives
    unsigned _statEventType[TR_MethodEvent::NumEvents];
    };
-} // namespace TR
 
-#endif // ifndef COMPILATIONCONTROLLER_INCL
+}; // namespace j9
+
+extern "C" {
+int32_t returnIprofilerState();
```
[the remark this answers is not in the document]
Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.hpp:244]: done

File: runtime/compiler/control/J9CompilationStrategy.cpp (hunk 3 above)
Comment (mpirvu) [reply runtime/compiler/control/J9CompilationStrategy.cpp:89]: In the old implementation this matrix was allocated by the compilation controller. What entity is allocating this data structure now?

Comment (jamesgua) [reply runtime/compiler/control/J9CompilationStrategy.cpp:89]: My bad: I missed it, now I add it back in [OMR](https://github.com/eclipse/omr/commit/32e3dd66941a8949e59e18367ac116cf6c6bc00a) guarded under J9_PROJECT_SPECIFIC
compiler/control/CompilationController.cpp after change at line 57

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11,jdk17 depends eclipse/omr#6884

**Comment by 0xdaryl** [comment]
@jamesgua : there are some build failures on Windows.  For example,

https://openj9-jenkins.osuosl.org/job/Build_JDK8_x86-64_windows_Personal/841/consoleText


```
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\omr\compiler\control\CompilationController.cpp(61): error C2039: 'compilation': is not a member of 'TR'
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\openj9\runtime\compiler\control/CompilationStrategy.hpp(28): note: see declaration of 'TR'
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\omr\compiler\control\CompilationController.cpp(61): error C2065: 'compilation': undeclared identifier
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\omr\compiler\control\CompilationController.cpp(70): error C2039: 'compilation': is not a member of 'TR'
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\openj9\runtime\compiler\control/CompilationStrategy.hpp(28): note: see declaration of 'TR'
[2023-03-14T15:01:51.965Z] F:\Users\jenkins\workspace\Build_JDK8_x86-64_windows_Personal\omr\compiler\control\CompilationController.cpp(70): error C2065: 'compilation': undeclared identifier
[2023-03-14T15:01:51.965Z] make[4]: *** [runtime/compiler/CMakeFiles/j9jit.dir/build.make:3547: runtime/compiler/CMakeFiles/j9jit.dir/49cc88ef8916a98e053ba6158b172f3d/omr/compiler/control/CompilationController.cpp.obj] Error 2
```

**Comment by jamesgua** [comment]
Update in [OMR](https://github.com/eclipse/omr/commit/311040155814436e542837b0bdbaa49b3dade86e) - TR::compilation corrected to TR::Compilation, pls rerun the build. Thx!

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11,jdk17 depends https://github.com/eclipse/omr/pull/6884

**Comment by 0xdaryl** [comment]
Jenkins test sanity zlinux jdk11 depends eclipse/omr#6884

**Comment by 0xdaryl** [comment]
Jenkins test sanity,sanity.openjdk all jdk8,jdk11,jdk17 depends https://github.com/eclipse/omr/pull/6884

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional plinux jdk8,jdk11,jdk17

**Comment by 0xdaryl** [comment]
Jenkins test sanity.openjdk plinux jdk8,jdk11

**Comment by 0xdaryl** [comment]
Lone Z failure appears to be #12582

**Comment by 0xdaryl** [comment]
Jenkins test sanity.openjdk plinux jdk8,jdk11 depends eclipse/omr#6884

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional plinux jdk17 depends eclipse/omr#6884

Looks like an infra issue.  Retrying.

File: runtime/compiler/build/files/common.mk [hunk 6]
```diff
@@ -380,6 +379,7 @@ JIT_PRODUCT_SOURCE_FILES+=\
     compiler/runtime/ValueProfiler.cpp \
     compiler/runtime/codertinit.cpp \
     compiler/runtime/emfloat.c \
+    omr/compiler/control/CompilationController.cpp \
```
Comment (0xdaryl) [inline runtime/compiler/build/files/common.mk:382]: Sorry, just discovered this.  You will also need to add the new file `compiler/control/J9CompilationStrategy.cpp` to this file.

As this does not affect builds in the open project I don't need to re-run all the tests and will just re-run a private build instead.

Comment (jamesgua) [reply runtime/compiler/build/files/common.mk:383]: Change updated, pls take another look.

**Comment by 0xdaryl** [comment]
For the record, all previous acceptance testing run on this PR with its OMR dependency was successful.

**Pull request merged**